chore(next.clayui.com): add more documentation about DropDown (#2601)
liferay/clay: next.clayui.com/content/docs/components/drop-down.mdx

	Check the <a href="https://liferay.design/lexicon">Lexicon</a> <a href="https://liferay.design/lexicon/core-components/dropdowns/">Dropdown Pattern</a> for a more in-depth look at the motivations and proper usage of this component.
</div>

<div class="clay-site-alert alert alert-warning">
	<strong class="lead">Warning</strong>

> 	Before you get started with this doc, read about <a href="/docs/foundations/composing.html" target="_blank">Clay's fundamentals of composition</a> to move on. The information below is assuming you have read about <a href="/docs/foundations/composing.html" target="_blank">Clay's compositional philosophy</a> and learned about the terms.

Author: That's a good point but I disagree a little with the idea that everyone can know the composition in React the vast majority maybe, but we have a lot of people who use Clay who don't know much about React or don't understand how it works. We should link our documents and make it clear what our proposal is being like, I think implicitly leaving may not be the best way to let people guess things...

I have the thought of creating a self serving documentation where people can find all the information. The idea of ​​adding this information to all documents is that many people can parachutes into each documentation and it is important that we pass this information on to people to understand our motives and eventually narrow the learning curve about using our lib.

> I slightly worry that by over communicating a compositional philosophy, we may scare people away in that they may think it is something additional they have to read through or adopt throughout their entire project.

Yeah that can be a problem, at least in the things I wrote #2585 it seems clear that we don't want to encourage but that's the way we found to better distribute Clay. I would not find it bad that the composition was to encourage actually I find the best way to be able to reuse the components in React in the application 🙂.

But we can narrow down this information in each documentation, since we have more than 30.

> 	Before you get started with this doc, read about <a href="/docs/foundations/composing.html" target="_blank">Clay's fundamentals of composition</a> to move on. The information below is assuming you have read about <a href="/docs/foundations/composing.html" target="_blank">Clay's compositional philosophy</a> and learned about the terms.

Reviewer: I think, like you said, a single document is likely the best way forward on describing this. Particularly if we can communicate "This is how we are distributing components, not necessarily how you have to." 

Overall just wanting to avoid any sort of overhead or headache for users when getting into clay. I would much rather have people play with examples rather than have them have to read an entire doc on composition. So the more we can promote examples, I think the better off we will be.

> 	Before you get started with this doc, read about <a href="/docs/foundations/composing.html" target="_blank">Clay's fundamentals of composition</a> to move on. The information below is assuming you have read about <a href="/docs/foundations/composing.html" target="_blank">Clay's compositional philosophy</a> and learned about the terms.

Author: > I think, like you said, a single document is likely the best way forward on describing this. Particularly if we can communicate "This is how we are distributing components, not necessarily how you have to."

Yeah, I agree with you about that. I'll take care to take a look at #2585 to make sure we're getting this message across.

> Overall just wanting to avoid any sort of overhead or headache for users when getting into clay. I would much rather have people play with examples rather than have them have to read an entire doc on composition. So the more we can promote examples, I think the better off we will be.

Yes, I don't want anyone reading this every time, but it would be essential to read it once before you start working with Clay, so it's the kind of document one will visit only once or when in doubt.

</div>

DropDown is delivered for composition offering flexibility for customization. DropDown's composition gives you the space to create more complex DropDowns, such as adding a form within DropDown, and creating fast DropDowns with the high-level `ClayDropDownWithItems` component that covers most Lexicon specification use cases.

<DropDown />

## Composing

You may want to use only the content of DropDown for other cases, such as mounting an `Autocomplete`, so use `DropDown.Menu` instead of `DropDown` which gives you the flexibility to only control content without a trigger.

<div class="clay-site-alert alert alert-warning">
	<code class="language-text">DropDown</code> consists of a <code class="language-text">DropDown.Menu</code> that controls the trigger and uses it as the basis for aligning DropDown content.
</div>

> #### Controlling Trigger

Author: I thought of putting a name other than the component just to get the ToC viewer's attention, `ClayDropDown.Menu` might not mean much to some.

> #### Controlling Trigger

Reviewer: I think the component name is helpful, but we could also call it something like "Using dropdown without a trigger." Mostly because I think "Controlling Trigger" isn't really helpful or descriptive what it is actually talking about.

> #### Controlling Trigger

Author: "Using dropdown without a trigger" looks better, I'll go with him.

You may want to create a trigger that is not necessarily in the same tree as DropDown, due to HTML markup issues, for these cases you can use `<ClayDropDown.Menu />`.

Using `<ClayDropDown.Menu />` allows you to better control the state of DropDown but you will have to deal with focus management, control expands and other details.

```jsx
import ClayDropDown from '@clayui/drop-down';
import React, {useState, useRef} from 'react';

const Menu = ({
    children,
    hasLeftSymbols,
    hasRightSymbols,
}) => {
    const triggerElementRef = useRef(null);
    const [expand, setExpand] = useState(false);
    const menuElementRef = useRef(null);

    const handleExpand = event => {
        // This is not ideal for allowing you to have more than
        // one trigger for the same content but it simulates the
        // advantages of controlling `DropDown.Menu`.
        triggerElementRef.current = event.target;

        setExpand(!expand);
    };

    return (

>         <div {...}>

Author: Hmm the intention was more to show that you would have more information on div, classNames etc... since I had written that you can control DropDown trigger using `DropDown.Menu` when you have tagging issues. So that was more to symbolize.

>         <div {...}>

Reviewer: I don't think users will think they _can't_ use other args on those elements, by leaving it out I think we communicate the same thing. Otherwise, if a user comes across and example without the `{...}` they may be unsure how to reconcile an example with and one without. So IMO I think its best to leave those out.

>         <div {...}>

Author: Yep, I think you're right.

            <div {...}>
                <button type="button" onClick={handleExpand}>Home</button>
            </div>
            <div {...}>
                <button type="button" onClick={handleExpand}>Product</button>
            </div>

            <ClayDropDown.Menu
                active={expand}
                alignElementRef={triggerElementRef}
                hasLeftSymbols={hasLeftSymbols}
                hasRightSymbols={hasRightSymbols}
                onSetActive={() => setExpand(!expand)}
                ref={menuElementRef}
            >
                {children}
            </ClayDropDown.Menu>
        </div>
    );
};
```

## DropDownWithItems

> Allows you to be able to create a faster DropDown, through its API you are able to create a Menu with groups of checkboxes and radios, links, buttons, search, caption... you will do the same as the components low-levels.

Reviewer: ```suggestion
Allows you to create a simple DropDown, through its API you are able to create a Menu with groups of checkboxes and radios, links, buttons, search, caption, etc.
```

Author: Maybe it would keep "would you do the same with low-level components" since it could do the same over there.

it's not fast 😱😞

Reviewer: > you will do the same as the components low-levels

I'm not sure this last portion is helpful in clarifying what DropDownWithItems is. My goal would be to use text and descriptions like this to help users understand strange caveats, but for general purpose of a component I think pointing to examples and API is generally the best for communicating purpose. Its often hard to describe similarities or differences between high/low and much easier to just point to examples of seeing how its different.

Author: Hmm yeah, really better to point to an example.
